### PR TITLE
ISSUE-86 Allow passing callbacks to ImapFuture

### PR DESCRIPTION
**Pull request by chibenwa**
This allow avoiding blocking in calling code to unwrap the future
and access the inner value. This empowers reactive/asynchronous/actor
systems integrations.

I chose the callback approach instead of using CompletableFuture in
order to minimize code changes.

## Description

Adds callbacks to ImapFuture

## Motivation and Context

Needs to call ImapFuture.get to do anything usefull with this library however it blocks and defeat any reactive/asynchronous/actor system use of this library.

## How Has This Been Tested?

Unit test. I will also ensure timely that this solves my original issue.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Major release (change is NOT backward compatible with prior release)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

Javadoc.

- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java [hunk 1]
```diff
@@ -101,6 +104,9 @@ public ImapResponseMapper() {
         if (valueType == ListStatusResult.class) {
             return (T) parser.parseToListStatusResult(content);
         }
+        if (valueType == EnableResult.class) {
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java:107]: same here. Please rebase again with master

File: core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java [hunk 2]
```diff
@@ -467,6 +473,39 @@ private IdResult parseToIdResult(@Nonnull final IMAPResponse[] ir) throws ImapAs
             return new IdResult(Collections.unmodifiableMap(serverParams));
         }
 
+        /**
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java:476]: same here. Please rebase again with master

File: core/src/test/java/com/yahoo/imapnio/async/data/EnableResultTest.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.yahoo.imapnio.async.data;
```
Comment (fansu) [inline core/src/test/java/com/yahoo/imapnio/async/data/EnableResultTest.java:1]: same here. Please rebase again with master

File: core/src/test/java/com/yahoo/imapnio/async/response/ImapResponseMapperTest.java [hunk 4]
```diff
@@ -1299,4 +1301,141 @@ public void testParseToSearchResultNotOK() throws IOException, ProtocolException
         Assert.assertNotNull(actual, "ImapAsyncClientException should occur.");
         Assert.assertEquals(actual.getFailureType(), FailureType.INVALID_INPUT, "Failure type mismatched.");
     }
+
```
Comment (fansu) [inline core/src/test/java/com/yahoo/imapnio/async/response/ImapResponseMapperTest.java:1304]: same here. Please rebase again with master

File: core/src/main/java/com/yahoo/imapnio/async/client/ImapFuture.java [hunk 5]
```diff
@@ -30,6 +31,46 @@
...
+    /** Called when the future is canceled. */
+    private final Runnable canceledCallback;
+
+    /**
+     * Base constructor.
+     * @param doneCallback Called when the future succeeds.
+     * @param exceptionCallback Called when the future fails.
+     * @param canceledCallback Called when the future is cancelled.
+     */
+    public ImapFuture(final Consumer<V> doneCallback, final Consumer<Exception> exceptionCallback, final Runnable canceledCallback) {
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/client/ImapFuture.java:47]: Please check out ImapFutureTest.java and add Unit Tests with use of your new ImapFuture(...) constructor.

You will need to test out the cases for done(), cancel() and those exception case

Comment (fansu) [reply core/src/main/java/com/yahoo/imapnio/async/client/ImapFuture.java:47]: any reason this need to be public? i would prefer this to be protected

File: core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java [hunk 6]
```diff
@@ -279,14 +280,37 @@ public void setDebugMode(@Nonnull final DebugMode newOption) {
...
+        }, new Runnable() {
+            @Override
+            public void run() {
+
+            }
+        });
+    }
+
+    @Override
+    public <T> ImapFuture<ImapAsyncResponse> execute(final ImapRequest command, final Consumer<ImapAsyncResponse> doneCallback,
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/internal/ImapAsyncSessionImpl.java:302]: Please check out ImapAsyncSessionImplTest.java and add Unit Tests with use of your new execute(...) function.

**Comment by chibenwa** [comment]
Hi,

I will take a look in the coming days.

I will be able to make those changes solely in ImapFuture by allowing registrations solely on this object.

When a callback is registered, we modify ImapFuture inner structured then immediately run it if the future is done. This makes it fully transparent from the existing interfaces and allows its use for the full asynchronous API, not just for the `exec` method.

Would this be an evolution of this work that would be welcomed?

Regards,

Benoit

File: core/src/main/java/com/yahoo/imapnio/async/client/ImapFuture.java (hunk 5 above)
Comment (chibenwa) [reply core/src/main/java/com/yahoo/imapnio/async/client/ImapFuture.java]: Constructor removed.

**Comment by chibenwa** [comment]
Hello, 

Any input on this topic dear imapnio maintainers?

Best regards,

Benoit TELLIER

**Review by fansu — APPROVED** [review]
@chibenwa Thanks for making these changes and supporting this library with enhancements. We are welcome for any contributions. Change looks good.

Please make change in **next PR** to upgrade pom version by minor to 4.3.8 in 
pom.xml
core/pom.xml
README.md

**Comment by chibenwa** [comment]
Cool!

As a side question, when to expect a `4.3.8` so that I can drop my fork?

Best regards,

Benoit

**Comment by fansu** [comment]
@chibenwa We just upgrade to 4.3.8, artifacts will be published within a day.

**Pull request merged**